### DISCUSSION TITLE
General ideas discussion thread

### DISCUSSION THREAD

**Discussion by doughsay** [ambry-app/ambry#571] [Ideas]
A starting place for discussing ideas on how to make Ambry more useful for more use-cases, and to reduce data-entry repetition.

**Comment by doughsay** [comment]
From an end-user perspective (an audiobook listener), we don't want to deal with "files".  The only thing that matters is the book you want to listen to and the metadata about the book (like Title, author(s), narrator(s), chapters, etc). I think Ambry is pretty solid from this perspective.

From a library maintainer perspective, we want to make it as easy and streamlined as possible to add new audiobooks to the server. This is where Ambry currently could use some help. Right now it's a pretty manual process that looks like this:

* first you need to create "people" records for the author(s) and narrator(s) of your audiobook.
  * authors can be imported from Audnexus, but it just takes the best/first-match and the data isn't always clean or correct
  * otherwise it's just manual data entry
* then you might create a "series" record if the book belongs to a series
  * this is manual data entry
* then you create a "book" record
  * You can import books by title (best/first-match, which is sometimes wrong) or by ASIN from the Audnexus API.
  * Descriptions are in HTML and Ambry requires markdown. Some "cleaning" is attempted on the HTML, but it's often bad/wrong
  * Not all fields are imported (e.g. publication date from Audnexus is for the audiobook recording, not for the first print publication of the book, and Ambry wants to be able to display the print publication date)
* once you have the authors, narrators, series and book records created, you can now create a "media" record
  * this is where you actuall upload files
  * this is manual data entry right now, nothing is auto imported (but there isn't much in this step, except for maybe the audiobook publication date)
  * saving and uploading the media record transcodes the uploaded "source" files into an output mp4 file with hls and mpeg-dash manifests. There are no choices, it's all hard-coded.
* Finally, once a media record exists, you can attempt to import chapters
  * Chapters can come from various sources, e.g. it can treat split mp3 files as each being a chapter, or it can attempt to extract chapters from m4b metadata, or it can import chapters from the Audnexus API and hope that they line up with your file.

As you can see, the process is not at all streamlined, and importing one audiobook alone can take 5-10 minutes of data entry and trial and error. I'd love to improve this, and so that's what this discussion thread is about: Brainstorming ideas on how to make this easier and more automated.

**Comment by doughsay** [comment]
I've been working on a complete re-imagining of the book upload process with an eye on automating as much as possible. I've written a GoodReads web-scraper and an API client for the undocumented but semi-public Audible API. There's still a lot to do, but I think this approach looks promising. The idea is to be able to upload file(s), type in a book title, and the entire rest of the metadata is auto-imported for you. Follow along here: https://github.com/ambry-app/ambry/pull/584

**Comment by RXWatcher** [comment]
Some thoughts:

1. I love that it uses postgres.
2. you should have shelves like [AudioBookShelf](https://www.audiobookshelf.org/). This would allow the admins to have 'shelves' of books on different topics.
3. Have a way to scan and import audiobooks. Look at ABS above for ideas. I have 186,000 (accurate number) audio books. Its not feasible to manually do anything with the importing of the books. ABS will scan a directory and use the metadata from google, audible, etc to automatically populate the books.

Why wouldnt I just use AudioBookShelf? Because you have postgres which solves a ton of perf issues for me.

Thanks for your time and effort in the project!
RX

**Reply by doughsay** [reply]
Great ideas, and yeah AudioBookShelf looks pretty cool (haven't tried it myself).

I'd love to be more feature compatible with AudioBookShelf, but the reality is that it's basically just me by myself working on Ambry and it looks like they have a huge community of contributors...  Shelves were on my list at one point to implement, just never got around to it.  The auto-scanning/importing is something I'd really love to figure out, it's just a very monumental task.
